app/services/dashboard/data.rb

```diff
@@ -1,17 +1,21 @@
 class Dashboard::Data
-  attr_reader :filter
+  attr_reader :filter, :settings
+
   # returns a status hash
   def self.status(filter = "")
     new(filter).report
   end
 
-  def initialize(filter = "")
+  def initialize(filter = "", settings = {})
     @filter = filter
     @report = {}
+    @settings = settings
   end
 
   def hosts
     @hosts ||= Host.authorized(:view_hosts, Host).search_for(filter).reorder('')
+    @hosts = @hosts.with_last_report_origin(settings[:origin]) if settings[:origin] && settings[:origin] != 'All'
```

Comment:
still getting undefined method [] on nil here app/services/dashboard/data.rb
```
 | ActionView::Template::Error: undefined method `[]' for nil:NilClass
 | /home/ares/Projekty/Zdrojaky/foreman/app/services/dashboard/data.rb:17:in `hosts'
 | /home/ares/Projekty/Zdrojaky/foreman/app/views/dashboard/_new_hosts_widget.html.erb:1:in `_237381f4a11f9d95c2d12bf2eee62699'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/template.rb:157:in `block in render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/notifications.rb:168:in `instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/template.rb:352:in `instrument_render_template'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/template.rb:155:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/deface-1.3.0/lib/deface/action_view_extensions.rb:41:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/partial_renderer.rb:342:in `block in render_partial'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/abstract_renderer.rb:42:in `block in instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `block in instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/abstract_renderer.rb:41:in `instrument'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/partial_renderer.rb:331:in `render_partial'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/partial_renderer.rb:310:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/renderer.rb:47:in `render_partial'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/renderer/renderer.rb:21:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/rendering.rb:103:in `_render_template'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/action_controller/metal/streaming.rb:217:in `_render_template'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionview-5.1.4/lib/action_view/rendering.rb:83:in `render_to_body'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/action_controller/metal/rendering.rb:52:in `render_to_body'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/action_controller/metal/renderers.rb:141:in `render_to_body'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/abstract_controller/rendering.rb:24:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/action_controller/metal/rendering.rb:36:in `render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/actionpack-5.1.4/lib/action_controller/metal/instrumentation.rb:44:in `block (2 levels) in render'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/activesupport-5.1.4/lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
 | /home/ares/.rbenv/versions/2.4.0/lib/ruby/2.4.0/benchmark.rb:308:in `realtime'
```

Reply:
All calls for settings now go through `Dashboard::Data` and when a widget has settings nil, it'll default now to a empty hash. Already existing Widgets with no settings should not cause a 500 anymore. 

To test it, clearing the settings from Widgets with this in `rails console`:
```
Widget.all.each {|w| w.data = {}; w.save! }
```